README
# Swibe
[![npm version](https://badge.fury.io/js/swibe.svg)](https://badge.fury.io/js/swibe)

A simple slide menu in JS

## About
- [Web](http://barcia.github.io/standarize)
- [Repository](https://github.com/barcia/standarize)
- [Issues](https://github.com/barcia/standarize/issues)
- [Changelog](https://github.com/barcia/swibe/blob/master/CHANGELOG)


## Getting Started

### 1. Load Swibe

#### Script
```html
<!-- Downloaded file -->
<script src="swibe.min.js"></script>

<!-- With UNPKG -->
<script src="https://unpkg.com/swibe"></script>
```

> [Download the last release](https://github.com/barcia/swibe/releases/latest)

#### Module
1. Install it with [npm](https://www.npmjs.com/package/swibe):

  ```js
  npm install swibe
  ```

2. Import *swibe.min.js* in your code.
  ```js
  import Swibe from 'swibe';
  ```



## 2. Start a instace
Start a Swibe menu instace with:

```js
new Swibe();
```


> See [config](https://github.com/barcia/swibe#configuration)


Add a **trigger** and a **menu** items
```html
<button id="swibe-menu-trigger">Menu</button>
```

```html
<nav id="swibe-menu">
  <ul>
    <li><a>Item</a></li>
    <li><a>Item</a></li>
    <li><a>Item</a></li>
    <li><a>Item</a></li>
  </ul>
</nav>
```

> You can see this example working in [demo.html](https://github.com/barcia/swibe/blob/master/docs/demo/index.html)




## Configuration

These are all settings with their default values
```js
new Swibe({
  menu: "swibe-menu",
  trigger: "swibe-menu-trigger",
  breakpoint: undefined, // px
  width: 300, // px
  time: 300, // ms
  shadowOpacity: 0.5, // Between 0 and 1
  aria: true,
  initialState: false,
  selfClose: false,
  zIndex: 999
});
```

* `menu`: _String_. The menu _**id**_.
* `trigger`: _String_. The _**id**_ of the element that opens the menu.
* `breakpoint`: _Number_. If you enter a value (in **pixels**), the menu will not work when the screen is wider.
* `width`: _Number_. The **width** of menu **in pixels**. Still, the menu never will be wider than the screen.
* `time`: _Number_. The animation time. In **miliseconds**.
* `shadowOpacity`: _Number_. The **opacity** of the shadow when the menu is open. Number **between 0 and 1**.
* `aria`: _Boolean_. By default, Swibe adds basic accessibility stuff to menu and trigger. Please, disable it only if you add these things manually.
* `initialState`: _Boolean_. Indicates if the menu is open or closed at startup.
* `selfClose`: _Boolean_. Indicates if the menu close by clicking on itself. This is useful when the links are links to page sections, load some Ajax, etc. and the page is not reloaded.
* `zIndex`: _Number_. `z-index` property of the _menu_. The _shadow_ has one point less..



### User Interface
* By default, _Swibe_ menu only add the **most basic styles for work**, but **not** for a good **UI**. Anyway, the demo has a **[basic stylesheet](https://github.com/barcia/swibe/blob/master/docs/demo/demo.css)** for a good UI, you can see it and use it.

* The three main elements of the Swibe menu: the _trigger_, the _menu_ and the _shadow_ has the data-attribute `data-state` with the value `true` or `false` if the menu is open or closed. You can use it with CSS or JavaScript.



## Accessibility
* Swibe menu add some basic **accessibility** stuff to _menu_ and _trigger_ elements. You can disable it with `aria: false`if you want add it manually.

* If the _trigger_ has **not text content** (For example, a _hamburguer_ icon), you should add it the `aria-label=''` attribute.

* I strongly recommend use a `<button>` as a _trigger_.

> If you can suggest improvements in a11y, are welcome.



## Credits
Developed by Iván Barcia
[Web](https://barcia.gal) · [Twitter](http://www.twitter.com/bartzia) · [GitHub](http://www.github.com/barcia)



## License
This project is under [MIT License](https://github.com/barcia/swibe/blob/master/LICENSE)


CHANGELOG
# Changelog
## [1.2.2](https://github.com/barcia/swibe/releases/tag/1.2.2) - 2019-12-30
### Updated
- Update NPM packages and dependencies

## [1.2.1](https://github.com/barcia/swibe/releases/tag/1.2.1) - 2019-02-21
### Fixed
- Small bug in IE11 that not return body scroll after close menu

## [1.2.0](https://github.com/barcia/swibe/releases/tag/1.2.0) - 2019-02-21
### Changed
- Changed [WebPack](https://webpack.js.org) bundler to [Rollup](https://rollupjs.org/)
- Remove extension from some markdown files
- Update README

## [1.1.0](https://github.com/barcia/swibe/releases/tag/1.1.0) - 2019-01-16
### Added
- Option to add Swibe as script
- WebPack
- Some config files

## [1.0.2](https://github.com/barcia/swibe/releases/tag/1.0.2) - 2019-01-11
### Changed
- Changed some package.json config

## [1.0.1](https://github.com/barcia/swibe/releases/tag/1.0.1) - 2019-01-09
### Changed
- Change folder structure

## [1.0.0-beta.5](https://github.com/barcia/swibe/releases/tag/1.0.0-beta.5) - 2018-12-19
### Added
- Add A11Y features
- Add _selfClose_ option
- Now the body scroll is disabled when the menu is open
- Add option to close the menu with the Escape key
### Changed
- Change config variable name from `status` to `initialStatus`

## [1.0.0-beta.4](https://github.com/barcia/swibe/releases/tag/1.0.0-beta.4) - 2018-12-18
### Added
- Add breakpoint feature
- Auto close if screen width is bigger than breakpoint

## [1.0.0-beta.3](https://github.com/barcia/swibe/releases/tag/1.0.0-beta.3) - 2018-12-14
### Changed
- Modify demo to load new JavaScript module
- Small tweaks


## [1.0.0-beta.2](https://github.com/barcia/swibe/releases/tag/1.0.0-beta.2) - 2018-12-14
### Changed
- Fix export module

## [1.0.0-beta.1](https://github.com/barcia/swibe/releases/tag/1.0.0-beta.1) - 2018-12-13
### Changed
- Prepare package for upload to [npmjs.com](https://npmjs.com)

## 0.4.0 - 2018-11-21
### Changed
- Rewrite Swibe.js
- Remove Sass file. Now all basic styles are added with JS


## 0.3.4 - 2018-03-16
### Changed
- Added CSS Custom Properties in CSS
- Now the CSS only use the ID, not a class.
- Rewrite all gulpfile.js

## 0.3.3 - 2018-01-12
### Changed
- Changed JS code to objects
- Now the Swibe script not need the 'defer' atribute.
- Many code improvements


## 0.3.2 - 2017-02-06
### Added
- New *responsive* mode: for change menu visibility depending of screen size
- Grunt files for development
- New minified JS file: `swibe.min.js`


## 0.3.1 - 2017-01-29
### Added
- CHANGELOG.md file
- Now the Swibe menu has releases :sunglasses:

### Changed
- New README.md from my own template
- Changed the name of the trigger id
- Changed the name of `.css` file


src/swibe.js
/**
 * Swibe 1.2.1
 * MIT License
 * https://github.com/barcia/swibe
 */

export default function Swibe(customConfig) {

	let state;

	const defaultConfig = {
		menu: "swibe-menu",
		trigger: "swibe-menu-trigger",
		breakpoint: undefined, // px
		width: 300, // px
		time: 300, // ms
		shadowOpacity: 0.5, // Between 0 and 1
		aria: true,
		initialState: false,
		selfClose: false,
		zIndex: 999
	};

	const config = Object.assign(defaultConfig, customConfig);


	// If all elements exists, init after the DOM is loaded.
	document.addEventListener('DOMContentLoaded', function () {
		swibe.load() && swibe.init();
	});


	const swibe = {
		load() {
			if (menu.get() && trigger.get()) {
				return true;
			} else {
				console.warn("Swibe items not loaded");
				return false;
			};
		},
		init() {
			menu.init();
			trigger.init();
			shadow.init();

			// Add event listeners
			if (config.breakpoint) {
				responsive.checkMediaQuery(responsive.mediaQuery);
				responsive.mediaQuery.addListener(responsive.checkMediaQuery)
			} else {
				this.addListeners();
			}

			// Set initial status
			config.initialState == false ? swibe.close() : swibe.open();
			return true;
		},
		addListeners() {
			trigger.element.addEventListener('click', swibe.toggle, false);
			document.addEventListener('touchstart', interaction.tap, false);
			document.addEventListener('touchmove', interaction.slide, false);
			document.addEventListener("keydown", interaction.keypress, false);
			config.selfClose && menu.element.addEventListener('click', swibe.close, false);
			return true;
		},
		removeListeners() {
			trigger.element.removeEventListener('click', swibe.toggle, false);
			document.removeEventListener('touchstart', interaction.tap, false);
			document.removeEventListener('touchmove', interaction.slide, false);
			document.removeEventListener("keydown", interaction.keypress, false);
			config.selfClose && menu.element.removeEventListener('click', swibe.close, false);
			return true;
		},
		open() {
			menu.open();
			trigger.open();
			shadow.open();
			document.body.style.overflow = 'hidden';
			state = true;
			return true;
		},
		close() {
			menu.close();
			trigger.close();
			shadow.close();
			document.body.style.overflow = ''; //IE11 Fallback
			document.body.style.overflow = null;
			state = false;
			return true;
		},
		toggle() {
			this.state ? swibe.close() : swibe.open();
			return true;
		}
	};


	const menu = {
		get() {
			this.element = document.getElementById(config.menu);
			return (this.element) ? true : false;
		},
		init() {
			this.element.style.position = 'fixed';
			this.element.style.top = 0;
			this.element.style.left = -config.width + 'px';
			this.element.style.width = config.width + 'px';
			this.element.style.maxWidth = 'calc(100vw - 40px)';
			this.element.style.height = '100vh';
			this.element.style.transition = 'left ease-in-out';
			this.element.style.transitionDuration = config.time + 'ms';
			this.element.style.willChange = 'left';
			this.element.style.overflowY = 'auto';
			this.element.style.zIndex = config.zIndex;
			config.aria && this.element.setAttribute('aria-labelledby', config.trigger);
			return true;
		},
		open() {
			this.element.style.left = 0;
			this.element.setAttribute('data-state', true);
			config.aria && this.element.setAttribute('aria-hidden', false);
			return true;
		},
		close() {
			this.element.style.left = -config.width + 'px';
			this.element.setAttribute('data-state', false);
			config.aria && this.element.setAttribute('aria-hidden', true);
			return true;
		}
	};


	const trigger = {
		get() {
			this.element = document.getElementById(config.trigger);
			return (this.element) ? true : false;
		},
		init() {
			config.aria && this.element.setAttribute('aria-controls', config.menu);
			config.aria && this.element.setAttribute('aria-haspopup', true);
			return true;
		},
		open() {
			this.element.setAttribute('data-state', true);
			config.aria && this.element.setAttribute('aria-expanded', true);
			return true;
		},
		close() {
			this.element.setAttribute('data-state', false);
			config.aria && this.element.setAttribute('aria-expanded', false);
			return true;
		},
	};


	const shadow = {
		init() {
			this.element = document.createElement('div');
			this.element.id = config.menu+'-shadow';
			this.element.style.position = 'fixed';
			this.element.style.backgroundColor = 'hsl(0, 0%, 0%)';
			this.element.style.top = 0;
			this.element.style.left = 0;
			this.element.style.zIndex = config.zIndex - 1;
			this.element.addEventListener('click', swibe.close);
			config.aria && this.element.setAttribute('aria-hidden', true);
			document.body.appendChild(this.element);
			return true;
		},
		open() {
			this.element.style.width = '100vw';
			this.element.style.height = '100vh';
			this.element.style.opacity = config.shadowOpacity;
			this.element.style.transition = 'opacity ease-in-out, height step-start, width step-start';
			this.element.style.transitionDuration = config.time + 'ms';
			this.element.setAttribute('data-state', true);
			return true;
		},
		close() {
			this.element.style.width = 0;
			this.element.style.height = 0;
			this.element.style.opacity = 0;
			this.element.style.transition = 'opacity ease-in-out, height step-end, width step-end';
			this.element.style.transitionDuration = config.time + 'ms';
			this.element.setAttribute('data-state', false);
			return true;
		}
	};


	const interaction = {
		tap(event) {
			let tap = event.touches[0];
			this.tapX = tap.clientX;
			this.tapY = tap.clientY;
			this.slideX = this.tapX;
			this.slideY = this.tapY;
			return true;
		},

		slide(event) {
			let slide = event.touches[0];
			this.slideX = slide.clientX;
			this.slideY = slide.clientY;

			if (!state) {
				if (interaction.gre.open(this.tapX, this.tapY, this.slideX, this.slideY)) {
					swibe.open();
					return true;
				} else {
					return false;
				};
			} else {
				if (interaction.gre.close(this.tapX, this.tapY, this.slideX, this.slideY)) {
					swibe.close();
					return true;
				} else {
					return false;
				};
			}
		},

		keypress(event) {
			if (event.key === "Escape" && state) {
				swibe.close();
			};
		},

		// Swibe Gesture Recognition Engine
		gre: {
			// Check if the conditions are right to open the menu
			open(tapX, tapY, slideX, slideY) {
				this.openCond1 = tapX < 20;
				this.openCond2 = slideX > tapX;

				if (this.openCond1 && this.openCond2) {
					return true;
				} else {
					return false;
				};
			},

			// Check if the conditions are right to close the menu
			close(tapX, tapY, slideX, slideY) {
				this.distX = slideX - tapX;
				this.distY = slideY - tapY;
				this.angle = ((Math.atan(this.distY / this.distX) * 180) / Math.PI) + 90;

				this.closeCond1 = slideX < tapX;
				this.closeCond2 = this.distY < 280;
				this.closeCond3 = this.distX < -30;
				this.closeCond4 = this.angle > (90 - 35);
				this.closeCond5 = this.angle < (90 + 35);

				if (this.closeCond1 && this.closeCond2 && this.closeCond3 && this.closeCond4 && this.closeCond5) {
					return true;
				} else {
					return false;
				};
			}
		}
	};


	const responsive = {
		mediaQuery: window.matchMedia('(max-width: ' + config.breakpoint + 'px)'),
		checkMediaQuery(event) {
			if (event.matches) {
				swibe.addListeners();
			} else {
				swibe.close();
				swibe.removeListeners();
			};
			return true;
		}
	};
};
